### ISSUE TITLE
prov/tcp fails to properly progress network messages in blocking mode

### ISSUE THREAD

**Issue by sydidelot** [ofiwg/libfabric#4247]
While testing the TCP provider in blocking mode, I figured out an issue which prevents the provider to correctly progress the incoming network messages.
I've found a workaround for that issue but I'm not 100% confident the workaround is correct. That's the reason I would appreciate some help from the community :-) 

In a nutshell, my network polling thread includes a code similar to the one below (this is pseudo-code, not the real one I'm using):

```
if (fi_trywait(fabric, cq_fid, 1) == FI_SUCCESS)
    epoll_wait(cq_fd, POLL_MAX_EVENTS, -1);
else 
    fi_cq_read(cq, &cq_entry, 1);
```

When using the TCP provider on a basic pingpong test (which is different that the one provided by fabtest), my code usually stops after some time because my network polling thread doesn't make progress anymore.
This issue seems to only happen when the pingpong involves RMA communications (fi_write/fi_read).
Also, I've tested with the new code from @vkrishna in PR #4239 and the bug is also reproducible there.
It's important to note that I've never experienced such issue with Verbs and socket providers, which makes me believe it's an issue with the TCP provider.

After digging into the problem, I realized that adding a `timeout` value to `epoll_wait()` in the pseudo-code above (`timeout` value greater or equal to 0) helps and my code is finally able to complete.

To me, it looks like an issue where the TCP provider allows to enter blocking while there are some network messages waiting for transmissions/receptions, and those messages don't modify the state of the file descriptors monitored by `epoll_wait()`. Or maybe incoming network events don't properly signal my polling thread blocked in `epoll_wait()`.

I had a quick look at the code and I might found something that might explain the issue.
I *think* the issue is that we should not allow `fi_trywait()` to return success when there are TX/RX messages waiting for progression. Right now, the `fi_trywait()` implementation in the TCP provider looks like this:

```
static int tcpx_try_func(void *util_ep)
{
	/* nothing to do here. When endpoints
	 have incoming data, cq drives progress*/
	return FI_SUCCESS;
}
```

This code always allow to enter blocking. Instead, I *think* the code should check if there are any TX/RX entries waiting for progression. 
If there is any, then `tcpx_try_func()` should probably return `-FI_EAGAIN` to prevent the code to enter blocking mode.

Here is what I suggest for the fix:
```
static int tcpx_try_func(void *util_ep)
{
	struct tcpx_ep *ep;
	ep = container_of(util_ep, struct tcpx_ep, util_ep);

	if (!slist_empty(&ep->tx_queue))
		return -FI_EAGAIN;

	if (!ep->cur_rx_entry)
		return -FI_EAGAIN;

	return FI_SUCCESS;
}
```

I've tested this simple patch with my pingpong testing code and it seems to have fixed the issue, but I'm not sure if I do this right or not.

Thanks in advance for your help on this issue!

**Comment by vkrishna** [comment]
Thanks for finding this. Yes I see that there is an issue here with pending TX messages. When we block on sock fds, we only wait for incoming messages events (FI_EPOLL_IN). When you have only pending TX messages, then blocking on fds would not help, and I believe that the try function should progress them. 

As per the receive part, we should be able to catch the fd events and progress RXs as we currently do.

**Comment by vkrishna** [comment]
Please go ahead. I think we should be fine without rx part below

>	if (!ep->cur_rx_entry)
>		return -FI_EAGAIN;

**Comment by sydidelot** [comment]
I've tested without the RX part you mentioned (TX part sill there), but my testing application still hangs. 
It look like both TX and RX changes are required to fix the progression issue.

**Comment by vkrishna** [comment]
I think this requires little more thought. It is possible that a send would come while app thread is waiting on blocking cq read. And assuming there is no rx message in progress, app would still hang since app thread won't wake up (as we are not signaling it in this case).

**Comment by sydidelot** [comment]
Oh Indeed, you are right -- we need to signal the app polling thread for every TX/RX queued into the lists. 
I'll rework my patch to address this issue as well.

**Comment by shefty** [comment]
I'm not sure the proposed fix is correct.  It doesn't allow blocking even in the case where the socket is unable to make progress.

Doesn't the tx_queue simply mean that there's application data ready to be sent?  If that's the case, we don't want to spin waiting for the socket to become ready.  Likewise, cur_rx_entry doesn't mean that there's data waiting to be read.

**Comment by shefty** [comment]
I'm not sure what the proper fix is.  :/

I agree with @vkrishna that we shouldn't need a receive side check.  If there's data waiting to be read on the socket, poll/select should return immediately.  Maybe there's a bug in how epoll is being used?

I *think* if we add EPOLLOUT when the tx queue is non-empty as you suggest would work for the transmit side.

**Comment by sydidelot** [comment]
@vkrishna addressed the issue in #4305. We can now close it.
Thanks again!
